### PR TITLE
Playlist for youtube plugins

### PR DESCRIPTION
**Pull request by nthforloop**
- the plgn saves playlist-id to a file
- a cron runs automatic downloads to location provided by config.
  ignore-erros - to continue download of next video in the playlist, 
  - even if download fails
  - conversion of video fails.
    in short keep moving.
- some times you need to re-initialize the plugin, like recreate the files, so edited the init method.

### CODE REVIEW

**Comment by sum12** [comment]
@Diptanshu8  please do look at the diff and let me know !!!

File: plugins/youtube.py [hunk 1]
```diff
@@ -86,6 +97,33 @@ def queue(data,what,param):
...
+    open(plgn.playlistsfile,'a').write('\n{0}'.format(playid))
+    return 'Saved playid {0}'.format(playid)
+
+@plgn.schedule(cron(hour=range(2,6)+range(10,18), minute=[0],second=[0]))
+@plgn.command('startplaylist')
+def continueplaylist(*args, **kwargs):
+    """ DONT USE THIS """
+    playids = [i for i in open(plgn.playlistsfile,'r').read().split('\n') if i ]
+    logger.debug(playids)
+    for i in  playids:
```
Comment (Diptanshu8) [inline plugins/youtube.py:115]: Have you tested that youtube-dl works for downloading a playlist on the basis of playlist id itself ? I could see that you are just grouping out the playlist_id with the help of regex matching in line 101. If the playlist_id is sufficient enough then go going so far .

**Comment by Diptanshu8** [comment]
@sum12 There are few things we need to take care about this PR
- What if someone just wants to download a particular youtube-playlist once and doesn't want to add a cron for it ?
- I was thinking that rather than setting up a cron and storing id of playlist, the better approach would be make the `link-donwloader` function identify the type of URL and if it is a playlist then download the contents to a seperate folder. Thereup, we could write a seperate plugin to trigger crons. This some 3rd user, say someone other than me or you doesn't need to play with the actual code and could directly set functionality(either to download a playlist once or to setup a cron) directly form slack!
- Another advantage of going by the approach of letting `link_downloader` identify the URL is that we can keep a tab of how many videos of a particular playlist have been downloaded successfully there by bringing in more transperancy and efficiency in our plugin's functioning. I have written a rough script in the past. Have a look [here](http://pastie.org/private/ugkv0upmzg4umsc4purqdw). It might not be that optimized but the main highlighting features are to be able to make seperate folders for every playlist and keep a count on the number of videos that are downloaded plus the starting number can be set as to resume downloading at the same video where the process faced an error earlier(this feature is important mainly when it comes to audio conversion. Once downloaded, youtube-dl extracts audio and deletes the downloaded one. Next time when we start the downloads again, then this process again downloads the video, converts it and then concludes that the audio was already present). Hope I was clear !

**Comment by sum12** [comment]
those are some nice observations there, 
-  [outtmpl](https://github.com/rg3/youtube-dl/#output-template) from youtube-dl lets you write some really nice output-templates. like `~/MyVideos/%(playlist)s/%(title)s.%(ext)s`, here it will automatically create a folder with the name of your playlist. 
- [TODO] Also youtube-dl automatically figures out if the video is a playlist or a song, downloads automatically. I wonder where it would place a video in case `outtmpl` a `playlist` argument in it.
- [TODO] for the cron, yes we should be able to remove a playlist_id from the save entries. But I guess thats not a big requirement write now.
- for the delete part, yes, agree that youtube-dl does redownload the video if the conversion fails, but for that I use the `-k` so that downloads are not deleted.
- [TODO] for the start from a different point in playlist, I had also employed the same approach as your in that script. But that not really great, if I delete the video form the playlist, or from the downloaded folder then the numbering is missed up. So for that problem I would recommend [download-archive](https://github.com/rg3/youtube-dl/#how-do-i-download-only-new-videos-from-a-playlist)

**Comment by Diptanshu8** [comment]
@sum12 
- The outtmpl settings you mentioned are just the things I wanted to stress on. Now, consider a scenario wherein link_downloader extracts info about the URL given to it with `download=False`(You can refer [here](http://pastie.org/private/ugkv0upmzg4umsc4purqdw) line 21). Once it has the info about the URL, we can modify the `outtmpl` accordingly. That gets our work done.
- The `-k` param is a good option but not that worthy as you might have noticed that youtube_dl donwloads chunks of videos and doesn't merge them. This may not be good. I would suggest another approach for this. If we keep a hash-file containing hashes of all the videos that have been downloaded and are present at the moment in the playlist directory(considering the posibility of some user removing just one video, so he wants to check that cron next time downloads it again). For this to happen please refer to line 19 to line 34 [here](http://pastie.org/private/ugkv0upmzg4umsc4purqdw). Thus, when a video(or its audio) is downloaded and kept in the respective playlist directory, then the next time cron runs, the plugin's `link-downloader` function checks if the hash is present(if not then directly downloads the required video/audio) in  the hash-file. If the hash is present in the hash-file but corresponding video is absent in the playlist directory, then the plugin downloads, else it moves on. Hope I could justify you with my clarification. This will also help us keeping a track of what videos are new to the playlist and download only those.

**Comment by sum12** [comment]
I guess i was not clear enough on the outtmpl thing. 

There is no need to extract name if playlist explicitly. Youtuble-dl automatically does the folder-creation stuff for you. 

And the mechanism of hash which you are suggesting is already available via the download-archive option.

**Comment by Diptanshu8** [comment]
That's great. The merge roll out may help me make corresponding changes in link_downloader or should I pull the branch and start working and testing?

**Pull request merged**